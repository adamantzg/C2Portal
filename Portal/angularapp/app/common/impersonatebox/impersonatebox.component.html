
  <div class="panel">    
    <div class="form form-inline">
      <div class="form-group">
          <label *ngIf="showLabel">Impersonate: </label>
          <input [(ngModel)]="customer_code"
                    [typeahead]="customers"                    
                    [typeaheadOptionsLimit]="25"
                    (typeaheadLoading)="changeTypeaheadLoading($event)"
                    [typeaheadMinLength]="3"                    
                    typeaheadOptionField="combined_name"
                    placeholder="Type customer code"
                    class="form-control">          
          <button *ngIf="showButton" class="btn btn-primary m-l-xs" (click)="clicked()" [disabled]="!customer_code">Impersonate</button>
          <button class="btn btn-danger" (click)="clear(true)" *ngIf="customer">Clear</button>
          <label *ngIf="customer">{{customer.code}} {{customer.name}} {{customer.address6}}</label>
          <span class="alert alert-danger" *ngIf="showWarning">Customer with given code not found</span>
          <span *ngIf="autoCompleteLoading">Loading customers...</span>
      </div>      
    </div>
  </div>

