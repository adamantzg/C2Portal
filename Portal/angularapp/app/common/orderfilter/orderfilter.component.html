<div class="row">
  <div class="col-sm-12 form-inline">
    <div class="form-group">
      <label class="m-r-xs">From:</label>
      <input type="text" bs-datepicker class="form-control m-r-xs" bsDatepicker [(bsValue)]="filter.from" name="date" [bsConfig]="bsConfig" required>
    </div>
    <div class="form-group">
        <label class="m-r-xs"> To:</label>
        <input type="text" bs-datepicker class="form-control m-r-xs" bsDatepicker [(bsValue)]="filter.to" name="date" [bsConfig]="bsConfig" required>
    </div>
    <div class="form-group">
      <label class="m-r-xs">{{searchLabel}}</label>
      <input type="text" [(ngModel)]="filter.searchText" class="form-control m-r-xs">
    </div>
    <button class="btn m-l-xs" (click)="applyFilter()">Go</button>
    <button class="btn m-l-xs" (click)="clearFilter()">Clear</button>
  </div>
  <!-- <div class="col-sm-12 form-inline">
    <div class="form-group">
      <label>Filter results:</label>
      <input type="text" [(ngModel)]="filter.searchText" class="form-control">
    </div>
  </div> -->
  
</div>
