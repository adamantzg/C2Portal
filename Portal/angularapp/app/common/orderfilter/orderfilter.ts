export class OrderFilter {

    constructor(from?: Date, to?: Date, searchText: string = '') {
        this.from = from;
        this.to = to;
        this.searchText = searchText;
    }

    from: Date;
    to: Date;
    searchText = '';
}
