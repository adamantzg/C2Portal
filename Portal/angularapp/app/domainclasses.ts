
export class Product {
    id: number;
    code: string;
    name: string;
    description: string;
    combined_name: string;
}

export class Customer {
    code: string;
    name: string;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    county: string;
    analysis_codes1: string;
    currency: string;
    combined_name: string;
}

export class User {
    id: number;
    name: string;
    lastname: string;
    username: string;
    password: string;
    email: string;
    customer_code: string;
    address: string;
    phone: string;
    token: string;
    customer: Customer;
    roles: Role[];
    locked: boolean;
    isTopAdmin: boolean;
    isBranchAdmin: boolean;
    activated: boolean;
    isInternal: boolean;
    lastLogin: Date;
    role: string;
    role_id: number;
    permissions: Permission[];

    constructor() {
        this.customer = new Customer();
    }
}

export class Role {
    id: number;
    name: string;
    permissions: Permission[];
    // strPermissions: string[];

    constructor(id: number, name?: string, permissions?: Permission[]) {
        this.id = id;
        this.name = name;
        this.permissions = permissions;
    }
}


export class Invoice {
    customer: string;
    dated: Date | null;
    kind: string;
    reference: string;
    item_no: string;
    due_date: Date | null;
    amount: number | null;
    open_indicator: string;
    customer_order_num: string;
    selected: boolean;
    postcode: string;
}

export enum RoleId {
    User = 1,
    BranchAdmin = 2,
    Admin = 3
}

export class Order {
    order_no: string; // Sage order number
    customer_order_no: string; // Customer Order number
    date_entered: Date; // Date order received and input to Sage
    invoice_date: Date; // Currently will be to indicate "Delivered" flag. When there is an invoice date then the order is delivered
    date_required: Date;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    statusText: string;
    status: number;
    customer?: string;
    country: string;
    audits: OrderAudit[];
    productStockData: StockData[];
    details: OrderDetail[];
    planned: boolean;
}

export class OrderDetail {
    product: string;
    order_qty: number;
    val: number;
    order_line_no: number;
}

export  enum OrderStatus {
    Received = 5,
    InProgress = 7,
    Delivered = 8,
    Cancelled = 9
    // Processed = 'Processed',
    // Confirmed = 'Confirmed',
    // Delivered = 'Delivered',
    // Cancelled = 'Cancelled',
}
export class OrderAudit {
    audit_key: string; // need to join to the Order_no field of opheadm
    character_value: string; // limited to 6 to infer Despatch Note printed
    audit_date: Date; // is the date used for Order Processed date.
}

export class StockData {
    quantity: number;
    product: Product;
    ship_date: Date | null;
    discontinued: boolean;
    madeToOrder: boolean;
    expiresThisYear: boolean;
}

export class Holiday {
    name: string;
    date: Date;
}
export class HolidayExtracted {
    name: string;
    year: number;
    month: number;
    date: number;
    fullDate: Date;
}
export class Permission {
    id: number;
    name: string;
    selected: boolean;

    constructor(id?: number, name?: string, selected?: boolean) {
        this.id = id;
        this.name = name;
        this.selected = selected;
    }
}

export enum PermissionId {
    ViewAccountDetails = 1,
    ViewHolidayAdministration = 2,
    ViewInvoiceHistory = 3,
    ViewOrderHistory = 4,
    ViewStockSearch = 5,
    ViewUserAdministration = 6,
    ViewRoleAdministration = 7
}

export class ClearanceStock {
    product_code: string;
    long_description: string;
    range: string;
    exvat_price: number | null;
    price: number | null;
    freestock: number | null;
    discount: number;
}
