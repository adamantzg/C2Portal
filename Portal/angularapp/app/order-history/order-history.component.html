<br/>
<div class="col-sm-12" *ngIf="user.isTopAdmin || user.isBranchAdmin">
  <div class="alert alert-danger" *ngIf="errorMessage">{{errorMessage}}</div>
  <app-impersonatebox *ngIf="showImpersonationBox()" (CustomerChange)="showForCustomer($event)"></app-impersonatebox>
</div>
<div class="col-sm-12">
  <div class="panel panel-default" *ngIf="loading">
    <div class="text-center"><h4>Loading...</h4></div>
  </div>
  <div class="panel panel-default" *ngIf="!loading">
    <app-block-ui></app-block-ui>
    <div class="panel-body">
      <app-orderfilter [filter]="filter" (filterApplied)="applyFilter()" (filterCleared)="clearFilter()"></app-orderfilter>
      <div class="row">
        <div class="col-sm-12">
          <div class="text-center">
              <pagination [itemsPerPage]="pageSize" [boundaryLinks]="true" [totalItems]="totals.orderCount" [maxSize]="20" 
              [(ngModel)]="page" (pageChanged)="pageChanged($event)"></pagination>
          </div>
          <app-grid [data]="filteredOrders" [definition]="gridDefinition" (ButtonClicked)="gridButtonClicked($event)" (SortChange)="onGridSortChange($event)" 
          [css]="'table table-striped'"></app-grid>
          <div class="text-center">
              <pagination [itemsPerPage]="pageSize" [boundaryLinks]="true" [totalItems]="totals.orderCount" [maxSize]="20" 
              [(ngModel)]="page" (pageChanged)="pageChanged($event)"></pagination>
          </div>          
        </div>
      </div>
    </div>
  </div>
</div>