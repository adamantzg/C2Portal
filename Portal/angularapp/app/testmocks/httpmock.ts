import { of } from 'rxjs';

export class HttpMock {
    url: string;
    data: any;
    params: any;
    options: any;
    postResult: any;
    getResult: any;


    post<T>(url, data, params) {
      this.data = data;
      this.url = url;
      this.params = params;
      return of(this.postResult);
    }

    get<T>(url, options) {
        this.url = url;
        this.options = options;
        return of(this.getResult);
    }

  }
