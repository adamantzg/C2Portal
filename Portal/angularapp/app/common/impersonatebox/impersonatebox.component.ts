import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { UserService } from '../../services/user.service';
import { Customer, RoleId } from '../../domainclasses';
import { Observable, Subject, Subscriber } from 'rxjs';

@Component({
  selector: 'app-impersonatebox',
  templateUrl: './impersonatebox.component.html',
  styleUrls: ['./impersonatebox.component.css']
})
export class ImpersonateboxComponent implements OnInit {

  constructor(private userService: UserService) { }

  customer_code: string;
  customer: Customer;

  /*@Input()
  public get Code() {
    return this.customer_code;
  }
  public set Code(value: string) {
    this.customer_code = value;
  }

  @Input()
  public get Customer() {
    return this.customerObj;
  }
  public set Customer(value: any)  {
    this.customerObj = value;
  }*/

  @Input()
  showLabel = false;

  @Input()
  showButton = true;

  /*@Output()
  buttonClicked = new EventEmitter();*/
  /*@Output()
  CodeChange = new EventEmitter<string>();*/
  @Output()
  CustomerChange = new EventEmitter<Customer>();
  showWarning = false;

  customers: Customer[] = [];
  autoCompleteLoading = false;

  ngOnInit() {
    const storage = localStorage.getItem('impersonated_customer');
    if (storage != null) {
      this.customer = JSON.parse(storage);
      this.CustomerChange.emit(this.customer);
      // this.CodeChange.emit(this.customerObj.code);
      this.customer_code = this.customer.code;
    } else {
      this.CustomerChange.emit(null);
    }
    this.customers = Observable.create((observer: any) => {
      // Runs on every search
      observer.next(this.customer_code);
    }).mergeMap((code: string) => this.userService.customerSearch(code, RoleId.User));

  }

  clicked() {
    // extract code
    const code = this.customer_code.split(' ')[0];
    this.userService.getCustomer(code).subscribe(c => {
      this.customer = c;
      this.CustomerChange.emit(this.customer);
      // this.CodeChange.emit(this.customer_code);
      if (c != null) {
        localStorage.setItem('impersonated_customer', JSON.stringify(c));
      } else {
        this.customer_code = '';
        this.showWarning = true;
        setTimeout(() => {
          this.showWarning = false;
        }, 1000 );
        localStorage.removeItem('impersonated_customer');
      }
    });
    // this.buttonClicked.emit();
  }

  clear() {
    this.customer_code = '';
    this.customer = null;
    localStorage.removeItem('impersonated_customer');
    this.CustomerChange.emit(null);
  }

  changeTypeaheadLoading(e: boolean) {
    this.autoCompleteLoading = e;
  }


}
