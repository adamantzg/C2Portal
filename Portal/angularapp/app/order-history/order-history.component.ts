import { Component, OnInit, AfterViewInit } from '@angular/core';
import { HeaderService } from '../../c2-flex/services/header.service';
import { UserService } from '../services/user.service';
import { User, Customer, Order } from '../domainclasses';
import { Sort, SortDirection, GridDefinition, GridColumn, ColumnDataType,
  GridColumnType, GridButtonEventData, GridButton } from '../../common_components/components';
import { OrderFilter } from '../common/orderfilter/orderfilter';
import * as moment from 'moment';
import { CustomerTotals, OrderTotals } from '../modelclasses';
import { OrderService } from '../services/order.service';
import { CommonService } from '../../c2-flex/services/common.service';
import { BsLocaleService, enGbLocale, BsModalService } from 'ngx-bootstrap';
import { defineLocale } from 'ngx-bootstrap/chronos';
import { Router } from '@angular/router';
import { OrderdetailmodalComponent } from './orderdetailmodal/orderdetailmodal.component';
import { InvoiceService } from '../services/invoice.service';
import { MessageboxService } from '../common/messagebox/messagebox.service';

@Component({
  selector: 'order-history',
  templateUrl: './order-history.component.html',
  styleUrls: ['./order-history.component.css']
})
export class OrderHistoryComponent implements OnInit {

  constructor(
    private userService: UserService,
    private orderService: OrderService,
    private commonService: CommonService,
    private _localeService: BsLocaleService,
    private router: Router,
    private bsModalService: BsModalService,
    private invoiceService: InvoiceService,
    private messageBoxService: MessageboxService,
    private headerService: HeaderService
  ) {


    this._localeService.use('en-gb');
    defineLocale('en-gb', enGbLocale);
    this.headerService.title = 'Order history';

   }

   user: User = this.userService.CurrentUser;
   errorMessage = '';
   customer_code = '';
   customer: Customer;
   impersonated: Customer;
   loading = false;
   dateFormat = 'dd/MM/yyyy';
   numFormat = '1.2-2';
   filter: OrderFilter = new OrderFilter();
   totals: OrderTotals = new OrderTotals();
   pageSize = 50;
   page = 1;
   lastpage = 1;
   impersonation = false;
   filteredOrders: Order[] = [];
   sortBy = 'DESC';
   showDesc = true;
   showAsc= false;


   gridDefinition = new GridDefinition([
    new GridColumn('C2 Order No', 'order_no', GridColumnType.Label, 'c2order', null, null, null, null, true ),
    new GridColumn('Customer Order Number', 'customer_order_no', GridColumnType.Label, 'customerorder', null, null, null, null, true ),
    new GridColumn('Status', 'statusText', GridColumnType.Label, 'status', null, null, null, null, true),
    new GridColumn('Order Date', 'date_entered', GridColumnType.Label, 'date_entered', null, null, null, null, true,
     ColumnDataType.Date, {format: this.dateFormat}),
    new GridColumn('Required Date', 'date_required', GridColumnType.Label, 'date_required', null, null, null, null, true,
    ColumnDataType.Date, {format: this.dateFormat} ),
    new GridColumn('buttons', '', GridColumnType.ButtonGroup, 'buttons', null, null, null, null, false, null, null, null, false,
    null,
    [
      new GridButton('Details', 'view', '', 'btn'),
      new GridButton('Status', 'order', '', 'btn'),
      new GridButton('POD', 'pod', '', 'btn')
    ])
  ]);

  showImpersonationBox() {
    return this.user.isTopAdmin || this.user.isBranchAdmin;
  }

  ngOnInit() {
    this.initFilter();
    this.gridDefinition.sort = new Sort(this.gridDefinition.columns.find(c => c.field === 'date_entered'), SortDirection.Desc);
    this.getOrderTotals(this.user.customer_code);
    this.gridDefinition.columnButtonVisibilityCallback = this.checkOrderButtonVisibility;
    if (!this.showImpersonationBox()) {
      this.getOrders(this.user.customer_code);
    }

  }




  showForCustomer(c: Customer) {
    this.page = 1;
    if (c != null) {
      this.customer_code = c.code;
      this.impersonation = true;
    } else {
      this.customer_code = '';
      this.impersonation = false;
    }
    this.applyFilter();
  }

  applyFilter() {
    const customer = this.impersonation ? this.customer_code : this.user.customer_code;
    this.getOrderTotals(customer);
    this.getOrders(customer);
  }

  pageChanged(e: any) {
    this.page = e.page;
    if (this.page !== this.lastpage) {
      this.getOrders(this.impersonation ? this.customer_code : this.user.customer_code);
      this.lastpage = this.page;
    }
  }

  gridButtonClicked($event: GridButtonEventData) {
    if ($event.name === 'view') {
      this.viewOrder($event.row);
    }else if ($event.name === 'order') {
      this.goToTracking($event.row);
    } else if ($event.name === 'pod') {
      this.showPod($event.row);
    }
  }

  onGridSortChange($event) {
    this.getOrders(this.impersonation ? this.customer_code : this.user.customer_code);
  }

  getOrders(customer: string) {
    this.loading = true;
    const limits = this.getPageLimits();
    const sortBy = this.sortBy;
    this.orderService.getOrders(customer, this.filter.from, this.filter.to, limits.from, limits.to,
      this.gridDefinition.sort.column.field, this.gridDefinition.sort.direction, this.filter.searchText).subscribe( (data) => {

      const orders: Order[] = data.result;
      orders.forEach(o => {
        if (o.date_entered != null) {
          o.date_entered = new Date(o.date_entered);
        }
        if (o.date_required != null) {
          o.date_required = new Date(o.date_required);
        }

      });
      this.filteredOrders = orders;
      this.loading = false;
    },
    err => {
      this.errorMessage = this.commonService.getError(err);
      this.loading = false;
    });
  }

  getPageLimits() {
    const result: any = {};
    result.from = (this.page - 1) * this.pageSize;
    result.to = result.from + this.pageSize - 1;
    return result;
  }

  viewOrder(o: Order) {
    this.orderService.getDetails(o.order_no).subscribe(data => {
      if (data != null) {
        o.details = data.result;
        let total = 0;
        o.details.forEach(d => {
          total += d.val;
        });
        const modal = this.bsModalService.show(OrderdetailmodalComponent);
        modal.content.order = o;
        modal.content.total = total;
      }
    });
  }

  goToTracking(o: Order) {
    this.router.navigate(['./c2/order-tracking'], {queryParams: {'order_no': o.order_no, 'return_to': 'orders'}});
  }

  getOrderTotals(customer: string) {
    this.loading = true;
    this.orderService.getOrderTotals(customer, this.filter.from, this.filter.to, this.filter.searchText).subscribe( data =>  {
      this.totals = data.result;
      this.loading = false;
    } ,
      err =>  {
        this.errorMessage = this.commonService.getError(err);
        this.loading = false;
      });
  }

  clearFilter() {
    this.initFilter();
    this.applyFilter();
  }

  initFilter() {
    this.filter.from = moment().subtract(6, 'month').toDate();
    this.filter.to = moment().toDate();
    this.filter.searchText = '';
  }

  checkOrderButtonVisibility(o: Order, name: string, buttonName: string) {
    if (buttonName === 'pod') {
      // tslint:disable-next-line:triple-equals
      return o.statusText === 'Invoiced' && o.planned != true;
    }
    return true;
  }

  showPod(o: Order) {
    /*const modal = this.bsModalService.show(PodModalComponent);
    modal.content.Invoice = i;*/

      const customer = this.impersonation ? this.customer_code : this.user.customer_code;
      this.invoiceService.checkInvoicePod(o.order_no, customer, o.customer_order_no, o.address6).subscribe((data: any) => {
        if (!data) {
          this.messageBoxService.
          openDialog('Request for POD sent, you will receive a copy within 2 working days from Customer Services');
        } else {
          let url = `/ShowOrderStatus.html`;
          //url += `?mod=POR&src=LOTS&sub=tracking&act=find&ref=${o.order_no.trim()}&pc=${o.address6.replace(' ', '')}`;
          window.open(url);
        }
      },
      err => {
        this.errorMessage = this.commonService.getError(err);
      });


  }

}
