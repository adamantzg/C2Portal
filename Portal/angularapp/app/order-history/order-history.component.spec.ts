import { OrderHistoryComponent } from './order-history.component';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ImpersonateboxComponent } from '../common/impersonatebox/impersonatebox.component';
import { UserService } from '../services/user.service';
import { HttpService } from '../services/http.service';
import { HttpClient } from '@angular/common/http';
import { HttpMock } from '../testmocks/httpmock';
import { OrderService } from '../services/order.service';
import { CommonService } from '../../c2-flex/services/common.service';
import { BlockUIService } from 'common_components/services/block-ui.service';
import { MessageboxService } from '../common/messagebox/messagebox.service';
import { InvoiceService } from '../services/invoice.service';
import { ModalModule, BsLocaleService, BsModalService, PaginationComponent,
    TypeaheadModule, BsDatepickerModule, PaginationConfig } from 'ngx-bootstrap';
import { RouterTestingModule } from '@angular/router/testing';
import { routes } from '../routestest';
import { BlockUiComponent } from '../blockui/block-ui.component';
import { OrderfilterComponent } from '../common/orderfilter/orderfilter.component';
import { CommonAppModule } from 'common_components/common.module';
import { FormsModule } from '@angular/forms';
import { User, Customer } from '../domainclasses';
import { By } from '@angular/platform-browser';
import { GridDefinition, GridColumn, SortDirection, GridButtonEventData } from 'common_components/components';

describe('order history', () => {
    let fixture: ComponentFixture<OrderHistoryComponent>;
    let component: OrderHistoryComponent;
    let orderService;
    let userService;

    setupTestBed({
        declarations: [OrderHistoryComponent, ImpersonateboxComponent, BlockUiComponent, OrderfilterComponent, PaginationComponent],
        providers: [UserService, HttpService, {provide: HttpClient, useClass: HttpMock},
            OrderService, CommonService, BlockUIService, MessageboxService, InvoiceService, BsLocaleService,
            BsModalService,  PaginationConfig],
        imports: [ModalModule.forRoot(), RouterTestingModule.withRoutes(routes), CommonAppModule,
            FormsModule, TypeaheadModule.forRoot(), BsDatepickerModule.forRoot()]
    });

    beforeEach(() => {
        orderService = TestBed.get(OrderService);
        userService = TestBed.get(UserService);
        const u = new User();
        u.id = 1;
        jest.spyOn(userService, 'CurrentUser', 'get').mockImplementation(() => u);
        fixture = TestBed.createComponent(OrderHistoryComponent);
        component = fixture.componentInstance;
        component.gridDefinition = new GridDefinition([]);
    });

    test('it should create component', () => {
        expect(component).toBeTruthy();
        expect(fixture).toMatchSnapshot();
    });

    test('impersonate box should show ', () => {
        component.user.isTopAdmin = true;
        jest.spyOn(component, 'getOrderTotals').mockImplementationOnce((code) => {});
        jest.spyOn(component, 'getOrders').mockImplementationOnce((code) => {});
        jest.spyOn(component, 'showForCustomer').mockImplementationOnce((code) => {});
        fixture.detectChanges();
        const impBox = fixture.debugElement.query(By.directive(ImpersonateboxComponent));
        expect(impBox).toBeTruthy();
    });

    test('impersonate box should not show', () => {

        jest.spyOn(component, 'getOrderTotals').mockImplementationOnce((code) => {});
        jest.spyOn(component, 'getOrders').mockImplementationOnce((code) => {});

        fixture.detectChanges();
        const impBox = fixture.debugElement.query(By.directive(ImpersonateboxComponent));
        expect(impBox).toBeNull();
    });

    test('ngOnInit', () => {
        jest.spyOn(component, 'getOrderTotals').mockImplementationOnce((code) => {});
        jest.spyOn(component, 'getOrders').mockImplementationOnce((code) => {});
        component.gridDefinition = new GridDefinition([new GridColumn('col', 'date_entered')]);
        component.ngOnInit();
        expect(component.gridDefinition.sort.column.title).toBe('col');
        expect(component.gridDefinition.sort.direction).toBe(SortDirection.Desc);
    });

    test('showForCustomer', () => {

        const fn1 = jest.spyOn(component, 'getOrderTotals').mockImplementation((code) => {});
        const fn2 = jest.spyOn(component, 'getOrders').mockImplementation((code) => {});
        component.user.customer_code = 'c1';
        let c = new Customer();
        c.code = 'c2';

        component.showForCustomer(c);
        expect(component.page).toBe(1);
        expect(component.impersonation).toBeTruthy();
        expect(fn1.mock.calls.length).toBe(1);
        expect(fn1.mock.calls[0][0]).toBe('c2');
        expect(fn2.mock.calls.length).toBe(1);
        expect(fn2.mock.calls[0][0]).toBe('c2');

        c = null;
        component.showForCustomer(c);
        expect(component.impersonation).toBeFalsy();
        expect(fn1.mock.calls.length).toBe(2);
        expect(fn1.mock.calls[1][0]).toBe('c1');
        expect(fn2.mock.calls.length).toBe(2);
        expect(fn2.mock.calls[1][0]).toBe('c1');

        fn1.mockReset();
        fn2.mockReset();

    });

    test('pageChanged', () => {
        const e = {page: 3};
        component.lastpage = 2;
        component.impersonation = true;
        component.customer_code = 'code';
        const fn1 = jest.spyOn(component, 'getOrders').mockImplementation((code) => {});
        component.pageChanged(e);
        expect(component.page).toBe(e.page);
        expect(fn1.mock.calls.length).toBe(1);
        expect(fn1.mock.calls[0][0]).toBe(component.customer_code);
        expect(component.lastpage).toBe(e.page);

        // NO impersonation
        component.lastpage = 2;
        component.impersonation = false;
        component.pageChanged(e);
        expect(fn1.mock.calls[1][0]).toBe(userService.CurrentUser.customer_code);

        fn1. mockReset();
    });

    test('gridButtonClicked', () => {
        const event = new GridButtonEventData('col', {reference: 'ref'});
        event.name = 'view';
        const fn1 = jest.spyOn(component, 'viewOrder').mockImplementationOnce(() => {});
        component.gridButtonClicked(event);
        expect(fn1.mock.calls.length).toBe(1);
        expect(fn1.mock.calls[0][0]).toBe(event.row);

        event.name = 'pod';
        const fn2 = jest.spyOn(component, 'showPod').mockImplementationOnce(() => {});
        component.gridButtonClicked(event);
        expect(fn2.mock.calls[0][0]).toBe(event.row);

        event.name = 'order';
        const fn3 = jest.spyOn(component, 'goToTracking').mockImplementationOnce(() => {});
        component.gridButtonClicked(event);
        expect(fn3.mock.calls[0][0]).toBe(event.row);

    });
});
