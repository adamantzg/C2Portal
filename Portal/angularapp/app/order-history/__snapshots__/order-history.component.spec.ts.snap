// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order history it should create component 1`] = `
<order-history
  _localeService={[Function BsLocaleService]}
  bsModalService={[Function BsModalService]}
  commonService={[Function CommonService]}
  customer_code=""
  dateFormat={[Function String]}
  errorMessage=""
  filter={[Function OrderFilter]}
  filteredOrders={[Function Array]}
  gridDefinition={[Function GridDefinition]}
  impersonation="false"
  invoiceService={[Function InvoiceService]}
  lastpage={[Function Number]}
  loading="false"
  messageBoxService={[Function MessageboxService]}
  numFormat={[Function String]}
  orderService={[Function OrderService]}
  page={[Function Number]}
  pageSize={[Function Number]}
  router={[Function Router]}
  showAsc="false"
  showDesc={[Function Boolean]}
  sortBy={[Function String]}
  totals={[Function OrderTotals]}
  user={[Function User]}
  userService={[Function UserService]}
>
  <br /><div
    class="col-sm-12"
  >
    
    
  </div>
</order-history>
`;
